### PR TITLE
802.15.4 Send Raw Syscall

### PR DESCRIPTION
**Pull request by tyler-potyondy**
### Pull Request Overview

Tock currently provides a syscall to support forming, securing, and sending 15.4 packets. This syscall allows user processes to provide the security level, destination address, key_id, etc and a payload. From this information, the 15.4 capsule then encrypts the payload (if necessary) and then forms the proper headers. Although this is helpful in most generic cases, it does not allow for user processes to fully form / encrypt 15.4 packets in userspace. 

In OpenThread's platform abstraction layer, OpenThread passes a fully formed 15.4 packet to the PAL function interfacing with the radio. This creates a challenge within libtock-c as there does not exist a syscall command for sending preformed packets. The current implementation would require the PAL libtock-c function deconstruct the fully formed 15.4 packet to retrieve and pass the fields into `ieee802154_send(..)`.

This PR adds a syscall command for direct sending (i.e. sending a packet fully formed by the user process). This is primarily accomplished through a wrapper `enum` `PendingTX` that abstracts the `pending_tx` as either `Direct` `Parse` or `Empty`. This allows the 15.4 capsule to differentiate between fully formed packets and packets requiring additional processing.

The direct send syscall also required modification to the `Frame` struct as `FrameInfo`  is not relevant to a `Direct` frame (beyond the length field). These nuances and explanations should be fully documented in the accompanying comments.

### Testing Strategy

This pull request was tested using the 15.4 test suite. This implementation does not break the existing 15.4 send syscall (that frames / forms the 15.4 packet). Moreover, a new 15.4 `radio_tx_direct` test was implemented.

Both test cases exhibited the expected behavior when observed in Wireshark (both in sending and edge cases of attempting to send a buffer greater than the 15.4 max buffer size).

### TODO or Help Wanted

When refactoring this, I had a comment / question regarding the  `perform_tx_sync` function. This function checks for the case when the pending_tx is Empty. In this case, the function returns without an error. I preserved this behavior in this PR, but wanted to draw attention to this. In my opinion, this seems to be the correct behavior, as not having a pending_tx does not imply that an error occurred when sending. I'm curious if others agree or think this should be handled differently.

### Documentation Updated

To my knowledge, there is not documentation on the 15.4 specific syscalls. Please let me know if there is somewhere I should update this documentation.

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by bradjc** [comment]
Does the term direct come from Thread? I think "raw" would be a more common term for ignoring the kernel 15.4 stack.

**Comment by tyler-potyondy** [comment]
> Does the term direct come from Thread? I think "raw" would be a more common term for ignoring the kernel 15.4 stack.

No, this was an arbitrary choice. I'll go ahead and update `Direct` with `Raw`.

File: capsules/extra/src/ieee802154/framer.rs [hunk 1]
```diff
@@ -99,7 +99,52 @@ use kernel::ErrorCode;
...
+
+/// This enum wraps the `FrameInfo` struct and allows each sending type
+/// (Parse or Direct) to only store the relevant information. In the
+/// case of a Direct send, the `FrameInfo` struct is irrelevant as the
+/// packet has been fully formed by the userprocess. For a direct send,
+/// we only require knowledge on the frame length. In the case of a
+/// Parse send, the wrapper provides all required frame header information.
+#[derive(Eq, PartialEq, Debug)]
+enum FrameInfoWrap {
+    Direct(usize),
```
Comment (bradjc) [inline capsules/extra/src/ieee802154/framer.rs:113]: What about here?

**Review by bradjc — DISMISSED** [review]
Looks good to me.

File: capsules/extra/src/ieee802154/driver.rs [hunk 2]
```diff
@@ -177,9 +177,41 @@ impl KeyDescriptor {
...
+}
+
+impl PendingTX {
+    /// Returns true if the PendingTX is not `Empty`
+    fn is_some(&self) -> bool {
+        match self {
+            PendingTX::Empty => false,
+            _ => true,
+        }
+    }
```
Comment (brghena) [inline capsules/extra/src/ieee802154/driver.rs:204]: I'm like 1% dubious about making our own `is_some()` method for something that isn't actually an Option. I don't feel strongly about it though, and maybe it's okay.

File: capsules/extra/src/ieee802154/framer.rs [hunk 3]
```diff
@@ -99,7 +99,52 @@ use kernel::ErrorCode;
...
+            FrameInfoWrap::Raw(_) => {
+                // This should never be called for a Raw send. The Framer should never
+                // require information other than the Frame length for a Raw send. This
+                // warrants a panic condition as fetching the `FrameInfo` struct for a
+                // Raw send is undefined behavior.
+                panic!("FrameInfoWrap::Raw called when expecting FrameInfoWrap::Parse")
+            }
+            FrameInfoWrap::Parse(info) => *info,
+        }
+    }
```
Comment (brghena) [inline capsules/extra/src/ieee802154/framer.rs:147]: Similarly dubious about reusing the `expect` name. A little more so here because the API differs as well.

**Review by brghena — APPROVED** [review]
Seems good to me. It makes sense to support this feature for now.

In a hypothetical future, you might want to separate the normal and raw interfaces and allow the board file to choose which it supports. That way apps can't do raw things unless the board wants to allow it. But given that boards are going to work through OpenThread and OpenThread will work through this, I expect that to be the path forward for all boards for now.

File: capsules/extra/src/ieee802154/driver.rs (hunk 2 above)
Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/driver.rs:204]: I will update this to be `is_empty` (and will update the logic accordingly)

File: capsules/extra/src/ieee802154/framer.rs (hunk 3 above)
Comment (tyler-potyondy) [reply capsules/extra/src/ieee802154/framer.rs:147]: I agree that this is misleading. I will update this naming to be `get_info`
